File: core/tracer/store.go
package tracer

import (
	"sync"

	"github.com/AleckDarcy/reload/core/log"
)

type Storage struct {
	traces map[TraceID]*traces
	lock   sync.RWMutex
}

type traces struct {
	traces map[UUID]*wrapper
}

type Refer struct {
	UUID UUID
}

type wrapper struct {
	refer *Refer
	trace *Trace
}

var Store = &Storage{
	traces: map[TraceID]*traces{},
}

func NewStore() *Storage {
	return &Storage{
		traces: map[TraceID]*traces{},
	}
}

func (s *Storage) CheckByContextMeta(meta *ContextMeta) bool {
	s.lock.RLock()
	ts, ok := s.traces[meta.traceID]
	if ok {
		_, ok = ts.traces[meta.uuid]
	}
	s.lock.RUnlock()

	return ok
}

func (s *Storage) GetByContextMeta(meta *ContextMeta) (*Trace, bool) {
	var t *Trace

	s.lock.RLock()
	ts, ok := s.traces[meta.traceID]
	if ok {
		var w *wrapper
		if w, ok = ts.traces[meta.uuid]; ok {
			t = w.trace.Copy()
		}
	}
	s.lock.RUnlock()

	return t, ok
}

func (s *Storage) SetByContextMeta(meta *ContextMeta, trace *Trace) {
	s.lock.Lock()
	if ts, ok := s.traces[meta.traceID]; ok {
		if w, ok := ts.traces[meta.uuid]; ok {
			merge(w.trace, trace)
		} else {
			ts.traces[meta.uuid] = &wrapper{trace: trace.Copy()}
		}
	} else {
		s.traces[meta.traceID] = &traces{traces: map[UUID]*wrapper{meta.uuid: {trace: trace.Copy()}}}
	}
	s.lock.Unlock()
}

func (s *Storage) UpdateFunctionByContextMeta(meta *ContextMeta, function func(*Trace)) (*Trace, bool) {
	var t *Trace

	s.lock.Lock()
	ts, ok := s.traces[meta.traceID]
	if ok {
		var w *wrapper
		if w, ok = ts.traces[meta.uuid]; ok {
			function(w.trace)

			t = w.trace.Copy()
		}
	}
	s.lock.Unlock()

	return t, ok
}

func (s *Storage) DeleteByContextMeta(meta *ContextMeta) bool {
	s.lock.Lock()
	ts, ok := s.traces[meta.traceID]
	if ok {
		if _, ok = ts.traces[meta.uuid]; ok {
			delete(ts.traces, meta.uuid)

			if len(ts.traces) == 0 {
				delete(s.traces, meta.traceID)
			}
		}
	}
	s.lock.Unlock()

	return ok
}

func merge(dst, src *Trace) {
	for dstI, srcI, dstLen, srcLen := 0, 0, len(dst.Records), len(src.Records); dstI < dstLen && srcI < srcLen; dstI++ {
		dstRecord, srcRecord := dst.Records[dstI], src.Records[srcI]
		// insert srcRecord between dst.Records[dstI-1] and dst.Records[dstI]
		if dstRecord.Timestamp > srcRecord.Timestamp {
			dst.Records = append(dst.Records[:dstI+1], dst.Records[dstI:]...)
			dst.Records[dstI] = srcRecord

			srcI++
			dstLen++
		} else if dstRecord.Timestamp == srcRecord.Timestamp {
			// deep equal
			if dstRecord.MessageName == srcRecord.MessageName && dstRecord.Type == srcRecord.Type {
				srcI++
			}
		} else {
			dstI++
		}
	}
}

/**
 * 3milebeach note:
 * Use following methods with caution.
 * Parameter ref is used for helping build context-propagation mechanism.
 * Usually used under asynchronous communication patterns.
 */

// REQUEST
func (s *Storage) DoRequest(meta *ContextMeta, trace *Trace, record *Record, ref *Refer) {
	trace = trace.Copy()
	trace.Records = nil
	trace.AppendRecord(record)

	// 3milebeach note:
	// ref should be nil when receiving requests
	s.lock.Lock()
	if ts, ok := s.traces[meta.traceID]; ok {
		ts.traces[record.Uuid] = &wrapper{refer: ref, trace: trace}
	} else {
		s.traces[meta.traceID] = &traces{traces: map[UUID]*wrapper{meta.uuid: {refer: ref, trace: trace}}}
	}

	if ref == nil {
		log.Trace.PrintlnWithCaller("%s write trace %d:%s", meta.server, meta.traceID, record.Uuid)
	} else {
		log.Trace.PrintlnWithCaller("%s write trace with refer %d:%s -> %s", meta.server, meta.traceID, record.Uuid, ref.UUID)
	}

	s.lock.Unlock()
}

// RESPONSE
func (s *Storage) DoResponse(meta *ContextMeta, trace *Trace, record *Record) {
	s.lock.Lock()
	if ts, ok := s.traces[meta.traceID]; ok {
		if w, ok := ts.traces[meta.uuid]; ok {
			merge(w.trace, trace)

			// 3milebeach note:
			// ref should be nil when sending responses
			if ref := w.refer; ref != nil { // report to corresponding request's upstream request
				if wRef, ok := ts.traces[ref.UUID]; ok {
					merge(wRef.trace, w.trace)
					*trace = *wRef.trace
					trace.AppendRecord(record)

					log.Trace.PrintlnWithCaller("%s write trace to refer %d:%s", meta.server, meta.traceID, record.Uuid)
				} else {
					log.Trace.PrintlnWithCaller("%s no ref uuid %d:%s", meta.server, meta.traceID, meta.uuid)
				}
			} else {
				*trace = *w.trace
				trace.AppendRecord(record)

				log.Trace.PrintlnWithCaller("%s write trace %d:%s", meta.server, meta.traceID, record.Uuid)
			}

			delete(ts.traces, meta.uuid)
			if len(ts.traces) == 0 {
				delete(s.traces, meta.traceID)
			}
		} else {
			log.Trace.PrintlnWithCaller("%s no uuid %d:%s", meta.server, meta.traceID, meta.uuid)
		}
	} else {
		log.Trace.PrintlnWithCaller("%s no trace %d", meta.server, meta.traceID)
	}

	s.lock.Unlock()
}

File: reload/python/grpc/Store.py
#! /usr/bin/python

import threading    # for locks

def merge(dst, src):
    dstI, srcI, dstLen, srcLen = 0, 0, len(dst.records), len(src.records)
    while dstI < dstLen and srcI < srcLen:
        dstRecord, srcRecord = dst.records[dstI], src.records[srcI]
        if dstRecord.timestamp > srcRecord.timestamp:
            #TODO: do this by indexing pythonic way instead of two for loops
            for j in range(0, dstI):
                dst.records.extend([dst.records[j]])
            for i in range(dstI + 1, len(dst.records)):
                dst.records.extend([dst.records[i]])
            dst.records[dstI] = srcRecord

            srcI += 1
            dstLen += 1
        elif dstRecord.timestamp == srcRecord.timestamp:
            if dstRecord.message_name == srcRecord.message_name and dstRecord.type == srcRecord.type:
                srcI += 1
        else:
            dstI += 1

        dstI += 1


class Store:
    def __init__(self):
        pass

    lock = threading.RLock()

    # hashmap ->
    # traceID: traces
    # traces ->
    # UUID: Trace
    store = dict()

    def CheckByContextMeta(self, meta):
        ok = False

        with self.lock:
            # check if meta.traceID exists in store and get traces dict
            traces = self.store.get(meta.traceID)
            if traces:
                # check if meta.uuid yields us a trace -> set ok true
                tmp = traces.get(meta.uuid)
                if tmp:
                    ok = True

        return ok

    # GUARD against trace being None (check return value)
    def GetByContextMeta(self, meta):
        ok = False
        t = None

        with self.lock:
            # check if meta.traceID exists in store and get traces dict
            traces = self.store.get(meta.traceID)
            if traces:
                # check if meta.uuid yields us a trace -> set ok true
                t = traces.get(meta.uuid)
                if t:
                    ok = True
                    # shallow copy in python 2.7
                    t = t[:]

        return t, ok

    def SetByContextMeta(self, meta, trace):
        with self.lock:
            # check if meta.traceID exists in store and get traces dict
            traces = self.store.get(meta.traceID)
            if traces:
                # check if meta.uuid yields us a trace -> set ok true
                t = traces.get(meta.uuid)
                if t:
                    merge(t, trace)
                else:
                    traces[meta.uuid] = trace
            else:
                tmpTrace = {meta.uuid: trace}
                self.store[meta.traceID] = tmpTrace

    def UpdateFunctionByContextMeta(self, meta, function):
        ok = False
        t = None

        with self.lock:
            # check if meta.traceID exists in store and get traces dict
            traces = self.store.get(meta.traceID)
            if traces:
                # check if meta.uuid yields us a trace -> set ok true
                t = traces.get(meta.uuid)
                if t:
                    ok = True

                    function(t)
                    # shallow copy in python 2.7
                    t = t[:]

        return t, ok

    def DeleteByContextMeta(self, meta):
        ok = False

        with self.lock:
            # check if meta.traceID exists in store and get traces dict
            traces = self.store.get(meta.traceID)
            if traces:
                # check if meta.uuid yields us a trace -> set ok true
                t = traces.get(meta.uuid)
                if t:
                    ok = True
                    del traces[meta.uuid]

                    if len(traces) == 0:
                        del self.store[meta.traceID]

        return ok

    # internal use
    def GetStore(self):
        return self.store
